bauble/src/value/mod.rs

```diff
@@ -0,0 +1,723 @@
+
+#[derive(Clone)]
+pub struct Attributes(pub IndexMap<Spanned<String>, Val>);
+
+#[derive(Clone)]
+/// A value with attributes
+pub struct Val {
+    pub value: Spanned<Value>,
+    pub attributes: Spanned<Attributes>,
+}
+
+pub type Ident = Spanned<String>;
+
+pub type Map = Vec<(Val, Val)>;
+
+pub type Fields = IndexMap<Ident, Val>;
+
+pub type Sequence = Vec<Val>;
+
+// type path.
+type TypePath = TypeInfo;
+
+type AssetPath = String;
+
+#[derive(Clone)]
+pub enum FieldsKind {
+    Unit,
+    Tuple(Sequence),
+    Struct(Fields),
+}
+
+impl FieldsKind {
+    pub fn variant_kind(&self) -> VariantKind {
+        match self {
+            FieldsKind::Unit => VariantKind::Path,
+            FieldsKind::Tuple(_) => VariantKind::Tuple,
+            FieldsKind::Struct(_) => VariantKind::Struct,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ValueKind {
+    Unit,
+    Num,
+    Bool,
+    Str,
+    Opt,
+    Ref,
+    Array,
+    Map,
+    Tuple,
+    Struct,
+    Enum,
+    BitFlags,
+    Raw,
+}
+
+#[derive(Clone)]
+pub enum Value {
+    Unit,
+    Num(Decimal),
+    Bool(bool),
+    Str(String),
+    Opt(Option<Box<Val>>),
+    // Fully resolved path.
+    Ref(AssetPath),
+
+    Array(Sequence),
+    Map(Map),
+
+    Tuple(Sequence),
+    Struct(Option<TypePath>, FieldsKind),
+    Enum(TypePath, Ident, FieldsKind),
+
+    BitFlags(Option<TypePath>, Vec<Ident>),
+
+    Raw(String),
+}
+
+impl Value {
+    pub fn kind(&self) -> ValueKind {
+        match self {
+            Value::Unit => ValueKind::Unit,
+            Value::Num(_) => ValueKind::Num,
+            Value::Bool(_) => ValueKind::Bool,
+            Value::Str(_) => ValueKind::Str,
+            Value::Ref(_) => ValueKind::Ref,
+            Value::Map(_) => ValueKind::Map,
+            Value::Struct { .. } => ValueKind::Struct,
+            Value::Tuple { .. } => ValueKind::Tuple,
+            Value::Array(_) => ValueKind::Array,
+            Value::BitFlags(_, _) => ValueKind::BitFlags,
+            Value::Raw(_) => ValueKind::Raw,
+            Value::Opt(_) => ValueKind::Opt,
+            Value::Enum(_, _, _) => ValueKind::Enum,
+        }
+    }
+
+    pub fn type_info(&self) -> Option<&TypeInfo> {
+        match self {
+            Self::Struct(type_info, _) | Self::BitFlags(type_info, _) => type_info.as_ref(),
+            Self::Enum(type_info, _, _) => Some(type_info),
+            _ => None,
+        }
+    }
+}
+
+pub struct Object {
+    pub object_path: AssetPath,
+    pub type_path: Option<TypePath>,
+    pub path: String,
+    pub value: Val,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ConversionError {
+    ModuleNotFound,
+    TypeNotFound,
+    ObjectNotFound,
+    AmbigousUse,
+    UnknownAttribute,
+    ExpectedIdent,
+    MissingNameAttribute,
+    ExpectedExpliciteType,
+    ExpectedEnum,
+    ExpectedUnit,
+    ExpectedTupleFields,
+    ExpectedFields,
+    ExpectedBitfield,
+    UnknownVariant,
+    NotAnEnum,
```

Comment:
What's the difference between this and `ExpectedEnum`?

Reply:
Both were unused so removed them.